// ngx-materialize/esm5/src/tooltip/tooltip.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, ElementRef, Input, Renderer } from '@angular/core';
var MzTooltipDirective = /** @class */ (function () {
    function MzTooltipDirective(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    MzTooltipDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initElements();
    };
    /**
     * @return {?}
     */
    MzTooltipDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.elementRef.nativeElement.getAttribute('type') === 'checkbox') {
            this.targetElement = $(this.elementRef.nativeElement).next('label');
        }
        this.initTooltip();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    MzTooltipDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (this.targetElement) {
            this.initTooltip();
        }
    };
    /**
     * @return {?}
     */
    MzTooltipDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.renderer.invokeElementMethod(this.targetElement, 'tooltip', ['remove']);
    };
    /**
     * @return {?}
     */
    MzTooltipDirective.prototype.initElements = /**
     * @return {?}
     */
    function () {
        this.targetElement = $(this.elementRef.nativeElement);
    };
    /**
     * @return {?}
     */
    MzTooltipDirective.prototype.initTooltip = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ tooltipOptions = {
            delay: isNaN(this.delay) || this.delay == null ? 350 : this.delay,
            html: this.html || false,
            position: this.position || 'bottom',
            tooltip: this.tooltip,
        };
        this.renderer.invokeElementMethod(this.targetElement, 'tooltip', [tooltipOptions]);
    };
    MzTooltipDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[mzTooltip], [mz-tooltip]',
                },] },
    ];
    /** @nocollapse */
    MzTooltipDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer, },
    ]; };
    MzTooltipDirective.propDecorators = {
        "delay": [{ type: Input },],
        "html": [{ type: Input },],
        "position": [{ type: Input },],
        "tooltip": [{ type: Input },],
    };
    return MzTooltipDirective;
}());
export { MzTooltipDirective };
function MzTooltipDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzTooltipDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzTooltipDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    MzTooltipDirective.propDecorators;
    /** @type {?} */
    MzTooltipDirective.prototype.delay;
    /** @type {?} */
    MzTooltipDirective.prototype.html;
    /** @type {?} */
    MzTooltipDirective.prototype.position;
    /** @type {?} */
    MzTooltipDirective.prototype.tooltip;
    /** @type {?} */
    MzTooltipDirective.prototype.targetElement;
    /** @type {?} */
    MzTooltipDirective.prototype.elementRef;
    /** @type {?} */
    MzTooltipDirective.prototype.renderer;
}
