// ngx-materialize/esm5/src/collection/collection.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef, Renderer2 } from '@angular/core';
var MzCollectionComponent = /** @class */ (function () {
    function MzCollectionComponent(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    MzCollectionComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initElements();
        this.initCollectionHeader();
    };
    /**
     * @return {?}
     */
    MzCollectionComponent.prototype.initElements = /**
     * @return {?}
     */
    function () {
        this.collectionElement = $(this.elementRef.nativeElement).find('.collection');
        this.collectionHeaderElement = $(this.elementRef.nativeElement).find('.collection-header');
    };
    /**
     * @return {?}
     */
    MzCollectionComponent.prototype.initCollectionHeader = /**
     * @return {?}
     */
    function () {
        if (this.collectionHeaderElement.length > 0) {
            this.renderer.addClass(this.collectionElement[0], 'with-header');
        }
    };
    MzCollectionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mz-collection',
                    template: "<div class=\"collection\">\n  <ng-content></ng-content>\n</div>",
                    styles: [""],
                },] },
    ];
    /** @nocollapse */
    MzCollectionComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
    ]; };
    return MzCollectionComponent;
}());
export { MzCollectionComponent };
function MzCollectionComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzCollectionComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzCollectionComponent.ctorParameters;
    /** @type {?} */
    MzCollectionComponent.prototype.collectionElement;
    /** @type {?} */
    MzCollectionComponent.prototype.collectionHeaderElement;
    /** @type {?} */
    MzCollectionComponent.prototype.elementRef;
    /** @type {?} */
    MzCollectionComponent.prototype.renderer;
}
