// ngx-materialize/esm5/src/tab/tab.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ContentChildren, ElementRef, Input, QueryList, ViewChild, } from '@angular/core';
import { MzTabItemComponent } from './tab-item/tab-item.component';
var MzTabComponent = /** @class */ (function () {
    function MzTabComponent() {
    }
    /**
     * @return {?}
     */
    MzTabComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.initTabs();
    };
    /**
     * @return {?}
     */
    MzTabComponent.prototype.initTabs = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ options = {
            onShow: this.onShow,
            responsiveThreshold: this.responsiveThreshold,
            swipeable: this.swipeable,
        };
        $(this.tabs.nativeElement).tabs(options);
    };
    /**
     * @param {?} tabItemId
     * @return {?}
     */
    MzTabComponent.prototype.selectTab = /**
     * @param {?} tabItemId
     * @return {?}
     */
    function (tabItemId) {
        $(this.tabs.nativeElement).tabs('select_tab', tabItemId);
    };
    MzTabComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mz-tab',
                    template: "<ul #tabs\n  class=\"tabs\"\n  [class.tabs-fixed-width]=\"fixedTabWidth\">\n  <li class=\"tab\" [class.disabled]=\"tabItem.disabled\" *ngFor=\"let tabItem of tabItems.toArray()\">\n    <a [class.active]=\"tabItem.active\"\n      href=\"{{ tabItem.href ? tabItem.href : '#' + tabItem.link }}\" target=\"{{ tabItem.target }}\">\n      {{ tabItem.label }}\n    </a>\n  </li>\n</ul>\n<div>\n  <ng-content select=\"mz-tab-item\"></ng-content>\n</div>",
                    styles: [""],
                },] },
    ];
    /** @nocollapse */
    MzTabComponent.propDecorators = {
        "fixedTabWidth": [{ type: Input },],
        "onShow": [{ type: Input },],
        "responsiveThreshold": [{ type: Input },],
        "swipeable": [{ type: Input },],
        "tabs": [{ type: ViewChild, args: ['tabs',] },],
        "tabItems": [{ type: ContentChildren, args: [MzTabItemComponent,] },],
    };
    return MzTabComponent;
}());
export { MzTabComponent };
function MzTabComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzTabComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzTabComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    MzTabComponent.propDecorators;
    /** @type {?} */
    MzTabComponent.prototype.fixedTabWidth;
    /** @type {?} */
    MzTabComponent.prototype.onShow;
    /** @type {?} */
    MzTabComponent.prototype.responsiveThreshold;
    /** @type {?} */
    MzTabComponent.prototype.swipeable;
    /** @type {?} */
    MzTabComponent.prototype.tabs;
    /** @type {?} */
    MzTabComponent.prototype.tabItems;
}
