// ngx-materialize/esm5/src/input/input-prefix/input-prefix.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, ElementRef, Renderer } from '@angular/core';
var MzInputPrefixDirective = /** @class */ (function () {
    function MzInputPrefixDirective(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    MzInputPrefixDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.renderer.setElementClass(this.elementRef.nativeElement, 'prefix', true);
    };
    MzInputPrefixDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'i[mzInputPrefix], i[mz-input-prefix]',
                },] },
    ];
    /** @nocollapse */
    MzInputPrefixDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer, },
    ]; };
    return MzInputPrefixDirective;
}());
export { MzInputPrefixDirective };
function MzInputPrefixDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzInputPrefixDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzInputPrefixDirective.ctorParameters;
    /** @type {?} */
    MzInputPrefixDirective.prototype.elementRef;
    /** @type {?} */
    MzInputPrefixDirective.prototype.renderer;
}
