// ngx-materialize/esm5/src/button/button.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, Renderer, } from '@angular/core';
import { HandlePropChanges } from '../shared/index';
var MzButtonDirective = /** @class */ (function (_super) {
    tslib_1.__extends(MzButtonDirective, _super);
    function MzButtonDirective(elementRef, renderer) {
        var _this = _super.call(this) || this;
        _this.elementRef = elementRef;
        _this.renderer = renderer;
        return _this;
    }
    /**
     * @return {?}
     */
    MzButtonDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initHandlers();
        this.initMaterialize();
        _super.prototype.executePropHandlers.call(this);
    };
    /**
     * @return {?}
     */
    MzButtonDirective.prototype.initHandlers = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.handlers = {
            disabled: function () { return _this.handleDisabled(); },
            flat: function () { return _this.handleFlat(); },
            float: function () { return _this.handleFloat(); },
            large: function () { return _this.handleLarge(); },
            noWaves: function () { return _this.handleNoWaves(); },
        };
    };
    /**
     * @return {?}
     */
    MzButtonDirective.prototype.initMaterialize = /**
     * @return {?}
     */
    function () {
        this.renderer.setElementClass(this.elementRef.nativeElement, 'btn', true);
    };
    /**
     * @return {?}
     */
    MzButtonDirective.prototype.handleDisabled = /**
     * @return {?}
     */
    function () {
        this.renderer.setElementClass(this.elementRef.nativeElement, 'disabled', this.disabled);
    };
    /**
     * @return {?}
     */
    MzButtonDirective.prototype.handleFlat = /**
     * @return {?}
     */
    function () {
        this.renderer.setElementClass(this.elementRef.nativeElement, 'btn', !this.flat);
        this.renderer.setElementClass(this.elementRef.nativeElement, 'btn-flat', this.flat);
    };
    /**
     * @return {?}
     */
    MzButtonDirective.prototype.handleFloat = /**
     * @return {?}
     */
    function () {
        this.renderer.setElementClass(this.elementRef.nativeElement, 'btn-floating', this.float);
    };
    /**
     * @return {?}
     */
    MzButtonDirective.prototype.handleLarge = /**
     * @return {?}
     */
    function () {
        this.renderer.setElementClass(this.elementRef.nativeElement, 'btn-large', this.large);
    };
    /**
     * @return {?}
     */
    MzButtonDirective.prototype.handleNoWaves = /**
     * @return {?}
     */
    function () {
        this.renderer.setElementClass(this.elementRef.nativeElement, 'waves-effect', !this.noWaves);
        if (!this.flat) {
            this.renderer.setElementClass(this.elementRef.nativeElement, 'waves-light', !this.noWaves);
        }
    };
    MzButtonDirective.decorators = [
        { type: Directive, args: [{
                    selector: "\n    a[mz-button],\n    a[mzButton],\n    button[mz-button],\n    button[mzButton]",
                },] },
    ];
    /** @nocollapse */
    MzButtonDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer, },
    ]; };
    MzButtonDirective.propDecorators = {
        "disabled": [{ type: Input },],
        "flat": [{ type: Input },],
        "float": [{ type: Input },],
        "large": [{ type: Input },],
        "noWaves": [{ type: Input },],
    };
    return MzButtonDirective;
}(HandlePropChanges));
export { MzButtonDirective };
function MzButtonDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzButtonDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzButtonDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    MzButtonDirective.propDecorators;
    /** @type {?} */
    MzButtonDirective.prototype.disabled;
    /** @type {?} */
    MzButtonDirective.prototype.flat;
    /** @type {?} */
    MzButtonDirective.prototype.float;
    /** @type {?} */
    MzButtonDirective.prototype.large;
    /** @type {?} */
    MzButtonDirective.prototype.noWaves;
    /** @type {?} */
    MzButtonDirective.prototype.elementRef;
    /** @type {?} */
    MzButtonDirective.prototype.renderer;
}
