// ngx-materialize/esm2015/src/validation/validation.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MzErrorMessageComponent } from './error-message/index';
import { MzValidationComponent } from './validation.component';
export class MzValidationModule {
}
MzValidationModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                declarations: [
                    MzErrorMessageComponent,
                    MzValidationComponent,
                ],
                entryComponents: [MzErrorMessageComponent],
                exports: [
                    MzErrorMessageComponent,
                    MzValidationComponent,
                ],
            },] },
];
function MzValidationModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzValidationModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzValidationModule.ctorParameters;
}
