// ngx-materialize/esm5/src/textarea/textarea-prefix/textarea-prefix.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, ElementRef, Renderer } from '@angular/core';
var MzTextareaPrefixDirective = /** @class */ (function () {
    function MzTextareaPrefixDirective(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    MzTextareaPrefixDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.renderer.setElementClass(this.elementRef.nativeElement, 'prefix', true);
    };
    MzTextareaPrefixDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'i[mzTextareaPrefix], i[mz-textarea-prefix]',
                },] },
    ];
    /** @nocollapse */
    MzTextareaPrefixDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer, },
    ]; };
    return MzTextareaPrefixDirective;
}());
export { MzTextareaPrefixDirective };
function MzTextareaPrefixDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzTextareaPrefixDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzTextareaPrefixDirective.ctorParameters;
    /** @type {?} */
    MzTextareaPrefixDirective.prototype.elementRef;
    /** @type {?} */
    MzTextareaPrefixDirective.prototype.renderer;
}
