// ngx-materialize/esm5/src/modal/modal.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Directive, ElementRef, EventEmitter, Input, Output, Renderer, ViewChild, } from '@angular/core';
import { HandlePropChanges } from '../shared/index';
var MzModalComponent = /** @class */ (function (_super) {
    tslib_1.__extends(MzModalComponent, _super);
    function MzModalComponent(renderer) {
        var _this = _super.call(this) || this;
        _this.renderer = renderer;
        _this.close = new EventEmitter();
        return _this;
    }
    /**
     * @return {?}
     */
    MzModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initHandlers();
        this.initElements();
        this.handleProperties();
    };
    /**
     * @return {?}
     */
    MzModalComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.initModal();
    };
    /**
     * @return {?}
     */
    MzModalComponent.prototype.initElements = /**
     * @return {?}
     */
    function () {
        this.modalElement = $(this.modalElementRef.nativeElement);
    };
    /**
     * @return {?}
     */
    MzModalComponent.prototype.initHandlers = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.handlers = {
            options: function () { return _this.handleOptions(); },
        };
    };
    /**
     * @return {?}
     */
    MzModalComponent.prototype.initModal = /**
     * @return {?}
     */
    function () {
        this.renderer.invokeElementMethod(this.modalElement, 'modal', [this.options]);
    };
    /**
     * @return {?}
     */
    MzModalComponent.prototype.handleProperties = /**
     * @return {?}
     */
    function () {
        _super.prototype.executePropHandlers.call(this);
    };
    /**
     * @return {?}
     */
    MzModalComponent.prototype.handleOptions = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // extend complete function to emit close event on callback return
        var /** @type {?} */ originalCompleteFn = this.options && this.options.complete || (function () { });
        this.options = Object.assign({}, this.options, {
            complete: function () {
                originalCompleteFn();
                _this.close.emit();
            },
        });
    };
    /**
     * @return {?}
     */
    MzModalComponent.prototype.openModal = /**
     * @return {?}
     */
    function () {
        this.renderer.invokeElementMethod(this.modalElement, 'modal', ['open']);
    };
    /**
     * @return {?}
     */
    MzModalComponent.prototype.closeModal = /**
     * @return {?}
     */
    function () {
        this.renderer.invokeElementMethod(this.modalElement, 'modal', ['close']);
    };
    MzModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mz-modal',
                    template: "<div #modal\n  class=\"modal\"\n  [class.modal-fixed-footer]=\"fixedFooter\"\n  [class.bottom-sheet]=\"bottomSheet\"\n  [class.modal-fullscreen]=\"fullscreen\"\n>\n  <div class=\"modal-content\">\n    <ng-content select=\"mz-modal-header\"></ng-content>\n    <div>\n      <ng-content select=\"mz-modal-content\"></ng-content>\n    </div>\n  </div>\n  <div class=\"modal-footer\">\n    <ng-content select=\"mz-modal-footer\"></ng-content>\n  </div>\n</div>\n",
                    styles: [".modal:not(.bottom-sheet).modal-fullscreen{top:12px!important;margin:0 auto;width:calc(100% - 24px);height:calc(100% - 24px);max-height:none}.modal.bottom-sheet.modal-fullscreen{height:100%;max-height:none}/deep/ mz-modal-header h5,/deep/ mz-modal-header h6{margin-top:0}"],
                },] },
    ];
    /** @nocollapse */
    MzModalComponent.ctorParameters = function () { return [
        { type: Renderer, },
    ]; };
    MzModalComponent.propDecorators = {
        "bottomSheet": [{ type: Input },],
        "fixedFooter": [{ type: Input },],
        "fullscreen": [{ type: Input },],
        "options": [{ type: Input },],
        "close": [{ type: Output },],
        "modalElementRef": [{ type: ViewChild, args: ['modal',] },],
    };
    return MzModalComponent;
}(HandlePropChanges));
export { MzModalComponent };
function MzModalComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzModalComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzModalComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    MzModalComponent.propDecorators;
    /** @type {?} */
    MzModalComponent.prototype.bottomSheet;
    /** @type {?} */
    MzModalComponent.prototype.fixedFooter;
    /** @type {?} */
    MzModalComponent.prototype.fullscreen;
    /** @type {?} */
    MzModalComponent.prototype.options;
    /** @type {?} */
    MzModalComponent.prototype.close;
    /** @type {?} */
    MzModalComponent.prototype.modalElementRef;
    /** @type {?} */
    MzModalComponent.prototype.modalElement;
    /** @type {?} */
    MzModalComponent.prototype.renderer;
}
var MzModalHeaderDirective = /** @class */ (function () {
    function MzModalHeaderDirective() {
    }
    MzModalHeaderDirective.decorators = [
        { type: Directive, args: [{ selector: 'mz-modal-header' },] },
    ];
    return MzModalHeaderDirective;
}());
export { MzModalHeaderDirective };
function MzModalHeaderDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzModalHeaderDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzModalHeaderDirective.ctorParameters;
}
var MzModalContentDirective = /** @class */ (function () {
    function MzModalContentDirective() {
    }
    MzModalContentDirective.decorators = [
        { type: Directive, args: [{ selector: 'mz-modal-content' },] },
    ];
    return MzModalContentDirective;
}());
export { MzModalContentDirective };
function MzModalContentDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzModalContentDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzModalContentDirective.ctorParameters;
}
var MzModalFooterDirective = /** @class */ (function () {
    function MzModalFooterDirective() {
    }
    MzModalFooterDirective.decorators = [
        { type: Directive, args: [{ selector: 'mz-modal-footer' },] },
    ];
    return MzModalFooterDirective;
}());
export { MzModalFooterDirective };
function MzModalFooterDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzModalFooterDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzModalFooterDirective.ctorParameters;
}
