// ngx-materialize/esm5/src/modal/modal.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { MzInjectionModule } from '../shared/injection/injection.module';
import { MzModalCloseDirective } from './modal-close/index';
import { MzModalComponent, MzModalContentDirective, MzModalFooterDirective, MzModalHeaderDirective, } from './modal.component';
import { MzModalService } from './services/index';
var MzModalModule = /** @class */ (function () {
    function MzModalModule() {
    }
    MzModalModule.decorators = [
        { type: NgModule, args: [{
                    imports: [MzInjectionModule],
                    declarations: [
                        MzModalCloseDirective,
                        MzModalComponent,
                        MzModalContentDirective,
                        MzModalFooterDirective,
                        MzModalHeaderDirective,
                    ],
                    exports: [
                        MzModalCloseDirective,
                        MzModalComponent,
                        MzModalContentDirective,
                        MzModalFooterDirective,
                        MzModalHeaderDirective,
                    ],
                    providers: [MzModalService],
                },] },
    ];
    return MzModalModule;
}());
export { MzModalModule };
function MzModalModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzModalModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzModalModule.ctorParameters;
}
