// ngx-materialize/esm5/src/collapsible/collapsible.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ContentChildren, ElementRef, Input, QueryList, Renderer, ViewChild, } from '@angular/core';
import { MzCollapsibleItemComponent } from './collapsible-item/collapsible-item.component';
var MzCollapsibleComponent = /** @class */ (function () {
    function MzCollapsibleComponent(renderer) {
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    MzCollapsibleComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.handleDataCollapsible();
        this.initCollapsible();
    };
    /**
     * @return {?}
     */
    MzCollapsibleComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        $(this.collapsible.nativeElement).collapsible('destroy');
    };
    /**
     * @param {?} collapsibleItemIndex
     * @return {?}
     */
    MzCollapsibleComponent.prototype.close = /**
     * @param {?} collapsibleItemIndex
     * @return {?}
     */
    function (collapsibleItemIndex) {
        $(this.collapsible.nativeElement).collapsible('close', collapsibleItemIndex);
    };
    /**
     * @return {?}
     */
    MzCollapsibleComponent.prototype.initCollapsible = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ options = {
            accordion: false,
            onClose: this.onClose,
            onOpen: this.onOpen,
        };
        $(this.collapsible.nativeElement).collapsible(options);
    };
    /**
     * @return {?}
     */
    MzCollapsibleComponent.prototype.handleDataCollapsible = /**
     * @return {?}
     */
    function () {
        if (this.mode) {
            this.renderer.setElementAttribute(this.collapsible.nativeElement, 'data-collapsible', this.mode);
        }
    };
    /**
     * @param {?} collapsibleItemIndex
     * @return {?}
     */
    MzCollapsibleComponent.prototype.open = /**
     * @param {?} collapsibleItemIndex
     * @return {?}
     */
    function (collapsibleItemIndex) {
        $(this.collapsible.nativeElement).collapsible('open', collapsibleItemIndex);
    };
    MzCollapsibleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mz-collapsible',
                    template: "<ul #collapsible\n  class=\"collapsible\"\n  [class.popout]=\"popout\"\n  [hidden]=\"items.length === 0\"\n>\n  <ng-content></ng-content>\n</ul>",
                    styles: ["/deep/ .collapsible-header{clear:both}/deep/ .collapsible-body{clear:both}"],
                },] },
    ];
    /** @nocollapse */
    MzCollapsibleComponent.ctorParameters = function () { return [
        { type: Renderer, },
    ]; };
    MzCollapsibleComponent.propDecorators = {
        "mode": [{ type: Input },],
        "onClose": [{ type: Input },],
        "onOpen": [{ type: Input },],
        "popout": [{ type: Input },],
        "collapsible": [{ type: ViewChild, args: ['collapsible',] },],
        "items": [{ type: ContentChildren, args: [MzCollapsibleItemComponent,] },],
    };
    return MzCollapsibleComponent;
}());
export { MzCollapsibleComponent };
function MzCollapsibleComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzCollapsibleComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzCollapsibleComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    MzCollapsibleComponent.propDecorators;
    /** @type {?} */
    MzCollapsibleComponent.prototype.mode;
    /** @type {?} */
    MzCollapsibleComponent.prototype.onClose;
    /** @type {?} */
    MzCollapsibleComponent.prototype.onOpen;
    /** @type {?} */
    MzCollapsibleComponent.prototype.popout;
    /** @type {?} */
    MzCollapsibleComponent.prototype.collapsible;
    /** @type {?} */
    MzCollapsibleComponent.prototype.items;
    /** @type {?} */
    MzCollapsibleComponent.prototype.renderer;
}
