// ngx-materialize/esm5/src/sidenav/sidenav-link/sidenav-link.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, ViewEncapsulation } from '@angular/core';
var MzSidenavLinkComponent = /** @class */ (function () {
    function MzSidenavLinkComponent() {
    }
    MzSidenavLinkComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mz-sidenav-link',
                    template: "<li\n  [class.active]=\"active\"\n>\n  <ng-content></ng-content>\n</li>\n",
                    styles: [":host /deep/ a[class*=mdi-]::before{color:rgba(0,0,0,.54);margin:-1px 32px 0 0;vertical-align:middle}:host /deep/ a i,:host /deep/ a i.material-icons,:host /deep/ a i[class*=mdi-]{margin-top:-1px}:host /deep/ a i.material-icons::before,:host /deep/ a i::before,:host /deep/ a i[class*=mdi-]::before{vertical-align:middle}"],
                    encapsulation: ViewEncapsulation.Emulated,
                },] },
    ];
    /** @nocollapse */
    MzSidenavLinkComponent.propDecorators = {
        "active": [{ type: Input },],
    };
    return MzSidenavLinkComponent;
}());
export { MzSidenavLinkComponent };
function MzSidenavLinkComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzSidenavLinkComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzSidenavLinkComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    MzSidenavLinkComponent.propDecorators;
    /** @type {?} */
    MzSidenavLinkComponent.prototype.active;
}
