// ngx-materialize/esm2015/src/tab/tab-item/tab-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, } from '@angular/core';
export class MzTabItemComponent {
    /**
     * @return {?}
     */
    get link() {
        return this.tabItemId ? this.tabItemId : this.label.replace(/[^a-zA-Z0-9]/g, '').toLowerCase();
    }
}
MzTabItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'mz-tab-item',
                template: `<div id="{{ link }}" class="{{ class }}">
  <ng-content></ng-content>
</div>`,
                styles: [``],
            },] },
];
/** @nocollapse */
MzTabItemComponent.propDecorators = {
    "active": [{ type: Input },],
    "class": [{ type: Input },],
    "disabled": [{ type: Input },],
    "href": [{ type: Input },],
    "label": [{ type: Input },],
    "tabItemId": [{ type: Input },],
    "target": [{ type: Input },],
};
function MzTabItemComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzTabItemComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzTabItemComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    MzTabItemComponent.propDecorators;
    /** @type {?} */
    MzTabItemComponent.prototype.active;
    /** @type {?} */
    MzTabItemComponent.prototype.class;
    /** @type {?} */
    MzTabItemComponent.prototype.disabled;
    /** @type {?} */
    MzTabItemComponent.prototype.href;
    /** @type {?} */
    MzTabItemComponent.prototype.label;
    /** @type {?} */
    MzTabItemComponent.prototype.tabItemId;
    /** @type {?} */
    MzTabItemComponent.prototype.target;
    /** @type {?} */
    MzTabItemComponent.prototype.tabs;
    /** @type {?} */
    MzTabItemComponent.prototype.liElement;
}
