// ngx-materialize/esm5/src/pagination/pagination.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { HandlePropChanges } from '../shared/index';
var MzPaginationComponent = /** @class */ (function (_super) {
    tslib_1.__extends(MzPaginationComponent, _super);
    function MzPaginationComponent() {
        var _this = _super.call(this) || this;
        _this.currentPage = 1;
        _this.enableFirstAndLastPageButtons = false;
        _this.maxPageButtons = 5;
        _this.pageChange = new EventEmitter();
        return _this;
    }
    Object.defineProperty(MzPaginationComponent.prototype, "totalPages", {
        get: /**
         * @return {?}
         */
        function () {
            return Math.ceil(this.totalItems / this.itemsPerPage);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    MzPaginationComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initHandlers();
        this.renderButtons();
    };
    /**
     * @param {?} pageNumber
     * @return {?}
     */
    MzPaginationComponent.prototype.changeCurrentPage = /**
     * @param {?} pageNumber
     * @return {?}
     */
    function (pageNumber) {
        this.currentPage = pageNumber;
        this.pageChange.emit(pageNumber);
        this.renderButtons();
    };
    /**
     * @return {?}
     */
    MzPaginationComponent.prototype.firstPage = /**
     * @return {?}
     */
    function () {
        this.changeCurrentPage(1);
    };
    /**
     * @return {?}
     */
    MzPaginationComponent.prototype.initHandlers = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.handlers = {
            currentPage: function () { return _this.renderButtons(); },
            itemsPerPage: function () { return _this.renderButtons(); },
            maxPageButtons: function () { return _this.renderButtons(); },
            totalItems: function () { return _this.renderButtons(); },
        };
    };
    /**
     * @return {?}
     */
    MzPaginationComponent.prototype.lastPage = /**
     * @return {?}
     */
    function () {
        this.changeCurrentPage(this.totalPages);
    };
    /**
     * @return {?}
     */
    MzPaginationComponent.prototype.nextPage = /**
     * @return {?}
     */
    function () {
        if (this.currentPage < this.totalPages) {
            var /** @type {?} */ nextPage = this.currentPage + 1;
            this.changeCurrentPage(nextPage);
        }
    };
    /**
     * @return {?}
     */
    MzPaginationComponent.prototype.previousPage = /**
     * @return {?}
     */
    function () {
        if (this.currentPage !== 1) {
            var /** @type {?} */ previousPage = this.currentPage - 1;
            this.changeCurrentPage(previousPage);
        }
    };
    /**
     * @return {?}
     */
    MzPaginationComponent.prototype.renderButtons = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ buttonsCount = Math.min(this.maxPageButtons, this.totalPages);
        var /** @type {?} */ maxPosition = this.totalPages - buttonsCount;
        var /** @type {?} */ halfButtons = Math.floor(buttonsCount / 2);
        var /** @type {?} */ hiddenPagesBefore = (this.currentPage - halfButtons);
        if (hiddenPagesBefore > maxPosition) {
            hiddenPagesBefore = maxPosition + 1;
        }
        var /** @type {?} */ from = Math.max(hiddenPagesBefore, 1);
        var /** @type {?} */ to = Math.min(this.totalPages, from + this.maxPageButtons - 1);
        this.pages = Array(buttonsCount).fill(0).map(function (x, i) { return from + i; });
        if (this.currentPage > this.totalPages) {
            this.currentPage = this.pages[0];
        }
    };
    MzPaginationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mz-pagination',
                    template: "\n  <ul class=\"pagination\">\n    <mz-pagination-page-button [disabled]=\"currentPage === 1\" *ngIf=\"enableFirstAndLastPageButtons\">\n      <a (click)=\"firstPage()\"><i mz-icon [icon]=\"'first_page'\"></i></a>\n    </mz-pagination-page-button>\n    <mz-pagination-page-button [disabled]=\"currentPage === 1\">\n      <a (click)=\"previousPage()\"><i mz-icon [icon]=\"'chevron_left'\"></i></a>\n    </mz-pagination-page-button>\n    <mz-pagination-page-button *ngFor=\"let page of pages\"\n      [active]=\"page === currentPage\"\n    >\n      <a (click)=\"changeCurrentPage(page)\">{{ page }}</a>\n    </mz-pagination-page-button>\n    <mz-pagination-page-button [disabled]=\"currentPage === totalPages\">\n      <a (click)=\"nextPage()\"><i mz-icon [icon]=\"'chevron_right'\"></i></a>\n    </mz-pagination-page-button>\n    <mz-pagination-page-button [disabled]=\"currentPage === totalPages\" *ngIf=\"enableFirstAndLastPageButtons\">\n      <a (click)=\"lastPage()\"><i mz-icon [icon]=\"'last_page'\"></i></a>\n    </mz-pagination-page-button>\n  </ul>",
                    styles: [""],
                },] },
    ];
    /** @nocollapse */
    MzPaginationComponent.ctorParameters = function () { return []; };
    MzPaginationComponent.propDecorators = {
        "currentPage": [{ type: Input },],
        "enableFirstAndLastPageButtons": [{ type: Input },],
        "itemsPerPage": [{ type: Input },],
        "maxPageButtons": [{ type: Input },],
        "totalItems": [{ type: Input },],
        "pageChange": [{ type: Output },],
    };
    return MzPaginationComponent;
}(HandlePropChanges));
export { MzPaginationComponent };
function MzPaginationComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MzPaginationComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MzPaginationComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    MzPaginationComponent.propDecorators;
    /** @type {?} */
    MzPaginationComponent.prototype.currentPage;
    /** @type {?} */
    MzPaginationComponent.prototype.enableFirstAndLastPageButtons;
    /** @type {?} */
    MzPaginationComponent.prototype.itemsPerPage;
    /** @type {?} */
    MzPaginationComponent.prototype.maxPageButtons;
    /** @type {?} */
    MzPaginationComponent.prototype.totalItems;
    /** @type {?} */
    MzPaginationComponent.prototype.pageChange;
    /** @type {?} */
    MzPaginationComponent.prototype.pages;
}
